Remove scipy dependencies from TPESampler

amylase opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->
Retry of #3560.

> SciPy requires relatively large storage and memory. It causes trouble in some constrained environments. E.g., This PR resolves https://github.com/optuna/optuna/issues/3035.

## Description of the changes
<!-- Describe the changes in this PR. -->
Now SciPy is only a test dependency. QMCSampler users may experience import error due to missing scipy after this change.

TPESampler time consumption is doubled. You can read the detailed results from the links below:
- this branch: https://github.com/amylase/optuna/actions/runs/3358589727
- master: https://github.com/amylase/optuna/actions/runs/3358590880

c-bata commented:
@contramundum53 @not522 Could you review this PR?

not522 commented on optuna/samplers/_tpe/_truncnorm.py:
```diff
@@ -0,0 +1,175 @@
+    return math.log1p(math.exp(log_p - log_q)) + log_q
+
+
+def _log_diff(log_p: float, log_q: float) -> float:
```
I think it is more readable to return `log(p - q)`.

not522 commented on tests/samplers_tests/tpe_tests/test_truncnorm.py:
```diff
@@ -0,0 +1,37 @@
+    sys.version_info < (3, 8, 0), reason="SciPy 1.9.2 is not supported in Python 3.7"
+)
+@pytest.mark.parametrize(
+    "a,b", [(-np.inf, np.inf), (-10, +10), (-1e-3, +1e-3), (10, 100), (-100, -10)]
```
Could you add smaller `a` and `b` values? For example, `(-1, 1)`.

amylase replied:
added (-1, +1)

not522 reviewed (commented):
Thank you for your PR! I added minor comments. I'll check more detail and the benchmark results.
Just to confirm, does this PR keep the behavior except for numerical error?

amylase commented:
Thank you for the review @not522! I have updated my code as per your comments. Please have a look.

> Just to confirm, does this PR keep the behavior except for numerical error?

Yes, my intention is to keep the TPESampler behavior unchanged.

amylase commented:
Speedtest benchmark results are as follows:

master
```
[  0.00%] · For Optuna commit a8799363 <master>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-fakeredis
[ 50.00%] ··· Running (optimize.OptimizeSuite.time_optimize--).
[100.00%] ··· optimize.OptimizeSuite.time_optimize                                                                    ok
[100.00%] ··· ============================= ============
                storage, sampler, n_trials
              ----------------------------- ------------
                  inmemory, random, 1000      250±10ms
                 inmemory, random, 10000     5.00±0.01s
                   inmemory, tpe, 1000       3.17±0.2s
                  inmemory, cmaes, 1000       738±30ms
                   sqlite, random, 1000      42.2±0.2s
               cached_sqlite, random, 1000   42.7±0.3s
                   redis, random, 1000       6.74±0.05s
                cached_redis, random, 1000   6.70±0.02s
              ============================= ============
```

remove-scipy-dependency
```
[  0.00%] · For Optuna commit 84876b75 <remove-scipy-dependency>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-fakeredis
[ 50.00%] ··· Running (optimize.OptimizeSuite.time_optimize--).
[100.00%] ··· optimize.OptimizeSuite.time_optimize                                                                    ok
[100.00%] ··· ============================= ============
                storage, sampler, n_trials
              ----------------------------- ------------
                  inmemory, random, 1000      234±3ms
                 inmemory, random, 10000     4.78±0.03s
                   inmemory, tpe, 1000       5.34±0.04s
                  inmemory, cmaes, 1000       691±20ms
                   sqlite, random, 1000      41.6±0.5s
               cached_sqlite, random, 1000   40.7±0.1s
                   redis, random, 1000       6.77±0.02s
                cached_redis, random, 1000   6.76±0.02s
              ============================= ============
```

contramundum53 commented:
@not522 @amylase Should we also test the case where `trunc_below` and `trunc_above` is equal?

c-bata commented on optuna/samplers/_tpe/_truncnorm.py:
```diff
@@ -0,0 +1,175 @@
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+
+# 2. Redistributions in binary form must reproduce the above
+#    copyright notice, this list of conditions and the following
+#    disclaimer in the documentation and/or other materials provided
+#    with the distribution.
```
Like the CuPy's documentation, could you add the license page in our documentation?
https://docs.cupy.dev/en/stable/license.html#scipy

amylase replied:
added a new doc page.

amylase commented:
Updated the branch.
- Added a testcase suggested by @contramundum53. 
- Added a third-party license page
- Use scipy for better performance if available

not522 commented:
It seems still slower than master even if `scipy` is installed.

amylase commented:
@not522 yeah. scipy implementation is actually slower than ours in my environment. So I decided to completely stop using scipy code in TPESampler.

amylase commented:
Updated the branch. The major updates are:

### Included the changes from #3985 (by @nzw0301 )
Since #3985 includes the dependency on SciPy, I cherry-picked these changes here. 
 
### Lazily import SciPy
In my previous change, SciPy was eagerly imported. I fixed it.

## Performance Notes
I used the following code and command to evaluate the performance impact of this PR.

#### mybench.py
```python
import optuna
from optuna.samplers import TPESampler


def objective(trial: optuna.Trial) -> float:
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_int("y", -100, 100)
    return x**2 + y**2


def main():
    study = optuna.create_study(sampler=TPESampler())
    study.optimize(objective, n_trials=1000)


if __name__ == "__main__":
    main()
```

```shell
$ python -m timeit "import mybench; mybench.main()"
```

### Results

master: 4.04s
nzw0301:resolve-zero-divide-tpe (#3985): 9.18s
amylase:remove-scipy-dependency w/ scipy: 9.43s
amylase:remove-scipy-dependency w/o scipy: 91.9s

Diff between master and #3985 is from sp.stats.truncnorm.cdf which is 20x slower than sp.special.erf. I think this is inevitable.
```python
>>> timeit.timeit("sp.stats.truncnorm.cdf(0, a=-1, b=1)", setup="import scipy as sp; sp.stats.truncnorm.cdf(0, -1, 1)", number=10000)
0.9007200490004834
>>> timeit.timeit("sp.special.erf(0)", setup="import scipy as sp; sp.special.erf(0)", number=10000)
0.005753725999966264
```

#3985 change have TPESampler use our implementations of special functions much more times than it did. This is why the current code is **10x slower without SciPy**.

contramundum53 commented:
I think we could call `sp.special.erf` three times to work around `truncnorm.cdf`.

amylase commented:
@contramundum53 the code actually does so after several nested function calls. the problem is that the number of function calls are huge (there were 141008396 `_ndtr` calls in benchmark code in [my previous comment](https://github.com/optuna/optuna/pull/4105#issuecomment-1319722308))

amylase commented:
Reduced redundant function calls. now the benchmark code takes *49.9sec*.

not522 commented:
I tried to speed up `truncnorm`. In my environment, this branch is 3x faster than the PR. https://github.com/optuna/optuna/commit/327fe598b418d2ed31a3ebf32d73ea21b9c7e3c9

- Use `np.frompyfunc` instead of `np.vectorize`
- Use `np.ndarray` as arguments instead of  `float`
- Set `maxsize` of `lru_cache` is 1000 (the default is 128)

contramundum53 commented:
How about implementing https://stackoverflow.com/questions/457408/is-there-an-easily-available-implementation-of-erf-for-python with numpy? Would that further increase speed?

amylase commented:
Added more optimizations.
* Included @not522's optimization
* Added our own `erf` implementation
  * Followed the @contramundum53's suggestion. The implementation they suggested was very fast but not accurate enough.
  * I decided to use the cite the implementation of FreeBSD's standard C library. It has very permissive license as stated in the top of `_erf.py`.

Now the benchmark code finishes in *15.6 sec*, while scipy version takes in *9.43 sec* as reported in my previous comment.

not522 commented on optuna/samplers/_tpe/parzen_estimator.py:
```diff
@@ -184,24 +169,34 @@ def log_pdf(self, samples_dict: Dict[str, np.ndarray]) -> np.ndarray:
+                    cdf_func = _ParzenEstimator._trunc_normal_cdf
+                    p_accept = cdf_func(high, mus, sigmas, low, high) - cdf_func(
+                        low, mus, sigmas, low, high
+                    )
+
                     upper_bound = np.minimum(samples + q / 2.0, high)
                     lower_bound = np.maximum(samples - q / 2.0, low)
-                    cdf = cdf_func(upper_bound[:, None], mus[None], sigmas[None]) - cdf_func(
-                        lower_bound[:, None], mus[None], sigmas[None]
-                    )
+                    cdf = cdf_func(
+                        upper_bound[:, None], mus[None], sigmas[None], low, high
+                    ) - cdf_func(lower_bound[:, None], mus[None], sigmas[None], low, high)
                     log_pdf = np.log(cdf + EPS) - np.log(p_accept + EPS)
```
I noticed that it can be rewritten as follows. With this change, the speed is comparable to using SciPy in my environment.

```python
log_gauss_mass = truncnorm._log_gauss_mass((lower_bound[:, None] - mus) / sigmas, (upper_bound[:, None] - mus) / sigmas)
log_p_accept = truncnorm._log_gauss_mass((low - mus) / sigmas, (high - mus) / sigmas)
log_pdf = log_gauss_mass - log_p_accept
```

amylase replied:
This is blasting fast! The speed was comparable to SciPy version in my environment too.

amylase commented:
Included @not522's optimization on `log_pdf` it is now comparable or slightly faster than scipy version.

not522 commented:
Thank you for your long-term efforts. It's almost LGTM.

1. Please remove `_trunc_normal_cdf` and `truncnorm.cdf`. They are no longer needed.
2. `_bisect` has room for improvement, but I think it's OK to merge.

Please merge the master branch. I will check that this PR does not change the behavior.

c-bata commented on docs/source/index.rst:
```diff
@@ -97,6 +97,8 @@ License
 
 MIT License (see `LICENSE <https://github.com/optuna/optuna/blob/master/LICENSE>`__).
 
+Optuna uses the codes from SciPy project (see :doc:`Third-party License <license_thirdparty>`).
```
nit:

```suggestion
Optuna uses the codes from SciPy and fdlibm projects (see :doc:`Third-party License <license_thirdparty>`).
```

contramundum53 commented on optuna/samplers/_tpe/_truncnorm.py:
```suggestion
def _ndtri_exp(z0: np.ndarray):
    C = 2 * math.sqrt(math.pi * 2)
    large = z0 > math.log(0.5)
    y = np.sqrt(-z0)
    y[large] = np.sqrt(-np.log(1 - np.exp(z0[large])))
    x = np.full_like(y, -1)
    for i in range(4):
        w = np.sqrt(-log_ndtr(x))
        x += C * (w - y) * w * np.exp(0.5 * x ** 2 - w ** 2)
    x[large] = -x[large]
    return x
```
[Possible followup] An algorithm based on Newton's method is simpler and should be a *lot* faster.

not522 reviewed (approved):
LGTM!

contramundum53 reviewed (approved):
LGTM!
Thanks for the PR!
